### ISSUE TITLE
Example code does not work with library

### ISSUE THREAD

**Issue by brandonsavage** [CuyZ/Valinor#39]
I am trying to instantiate an object using this library, and using `DateTimeInterface` type hints as described in the example in the README. However, I get a CannotMapObject exception when I do. Does this library not yet support using a `DateTimeInterface` type hint for date times? Thanks.

**Comment by romm** [comment]
Hi @brandonsavage, sorry for the lack of documentation about this, I will take some time to expand on it.

What is the input value that you're trying to map to a DateTime? The library will try different format, see the test case below:

https://github.com/CuyZ/Valinor/blob/e5ccbe201b8767066e0e0510e606f9492f3270f1/tests/Integration/Mapping/Type/DateTimeMappingTest.php#L17-L44

In case the format is not common, you can use the syntax `['datetime' => 'foo', 'format' => 'some date format']`, please tell me if it can solve your issue. 🙂

**Comment by brandonsavage** [comment]
Thanks for getting back to me so quick. I was actually reviewing that test when your comment showed up. The date is a Postgres `datetime` column with precision (which is the way they come).  I’ll have to do some more research to see how to handle it and format it correctly. I may have to format it myself on its way out of the database. 

Thanks for taking a look at my issue. I think there’s a way to solve this. I assume that if it can’t determine a format it doesn’t try at all?

**Comment by romm** [comment]
The library could give a way to define some global custom format that would be tested automatically during mapping, I'll have to do some research, could look like:

```php
(new \CuyZ\Valinor\MapperBuilder())
    ->mapper()
    ->withDateTimeFormat('some custom format') // something handled by https://www.php.net/manual/en/datetime.format.php#refsect1-datetime.format-parameters
    ->map(SomeClass::class, [ /* ... */]);
```

Please let me know what is the exact format returned by these column, to see if we can do something out of the box. 🙂

**Comment by brandonsavage** [comment]
It appears the format returned by the columns is:

`Y-m-d H:i:s.u`

Note the precision on the seconds. It doesn't like the precision. Transforming the data into the format of an array specifying the datetime and the format did the trick.

**Comment by brandonsavage** [comment]
For what it's worth, MySQL outputs the time as a timestamp in the following format:

`YYYY-MM-DD hh:mm:ss`

It might make sense to support this format for both MySQL and PgSQL just to help developers out.

In addition, it would appear the dates and times in the example in the README do not gracefully work, either. The date is `"date": "1957-07-23 13:37:42",` but I wasn't able to find a format that would work for that in the code. Did I miss it?

I'm happy to do the work of submitting a patch, BTW. Working on one now.

**Comment by brandonsavage** [comment]
This was resolved by #39.
